docs/usage/use-built-in-abstract-repository.md

```diff
@@ -7,6 +7,336 @@ nav_order: 5
+{
+    public HeroByNameContainsFilterSpec(string name)
+    {
+        if (!string.IsNullOrEmpty(name))
+        {
+            Query.Where(h => h.Name.Contains(name));
+        }
+    }
+}
+
+public class HeroService
+{
+    private readonly YourRepository<Hero> repository;
+
+    public HeroService(YourRepository<Hero> repository)
+    {
+        this.repository = repository;
+    }
+
+    public async Task<List<Hero>> GetHeroesFilteredByName(string name)
+    {
+        var spec = new HeroByNameContainsFilterSpec(name);
+
+        return await repository.ListAsync(spec);
+    }
+}
+
+public class CustomerByNameContainsFilterSpec : Specification<Customer>
+{
+    public CustomerByNameContainsFilterSpec(string name)
+    {
+        if (!string.IsNullOrEmpty(name))
+        {
+            Query.Where(c => c.Name.Contains(name));
+        }
+    }
+}
+
+public class CustomerService
+{
+    private readonly YourRepository<Customer> repository;
+
+    public CustomerService(YourRepository<Customer> repository)
+    {
+        this.repository = repository;
+    }
+
+    public async Task<List<Customer>> GetCustomersFilteredByName(string name)
+    {
+        var spec = new CustomerByNameContainsFilterSpec(name);
+
+        return await repository.ListAsync(spec);
+    }
+}
+```
+
+## Features of `RepositoryBase<T>`
+
+The section above introduced using `RepositoryBase<T>` to provide similar functionality across two entities and their services. This section aims to go into more detail about the methods made available by `RepositoryBase<T>` and provide some examples of their usages. Continuing with the HeroService example, it is possible to create heroes as follows using the `AddAsync` method. The `SaveChangesAsync` method exposes the underlying DbContext method of the same name to persist changes to the database.
+
+```csharp
+public async Task<Hero> Create(string name, string superPower, bool isAlive, bool isAvenger)
+{
+    var hero = new Hero(name, superPower, isAlive, isAvenger);
+
+    await repository.AddAsync(hero);
+
+    await respository.SaveChangesAsync();
+
+    return hero;
+}
+```
+
+Now that a Hero has been created, it's possible to retrieve that Hero using either the Hero's Id or by using a Specification. Note that since the `HeroByNameSpec` returns a single Hero entity, the Specification inherits the interface `ISingleResultSpecification` which `GetBySpecAsync` uses to constrain the return type to a single Entity.
+
+```csharp
+public class HeroByNameSpec : Specification<Hero>, ISingleResultSpecification
+{
+    public HeroByNameSpec(string name)
+    {
+        if (!string.IsNullOrEmpty(name))
+        {
+            Query.Where(h => h.Name == name);
+        }
+    }
+}
+
+public async Task<Hero> GetById(int id)
+{
+    return await repository.GetByIdAsync(id);
+}
+
+public async Task<Hero> GetByName(string name)
+{
+    var spec = new HeroByNameSpec(name);
+
+    return await repository.GetBySpecAsync(spec);
```

Comment:
At this point in reading this I want to know what happens if there are 2 heroes with this same name...

Reply:
described how `GetBySpecAsync` would handle this above